MultiBuild/BlueprintManager.cs

```diff
@@ -554,124 +523,67 @@ public static List<BuildPreview> Paste(Vector3 targetPos, float yaw, bool pasteI
-                BeltCopy belt = data.copiedBelts[i];
-                Vector2 newRelative = belt.cursorRelativePos.Rotate(yawRad, belt.originalSegmentCount);
-                Vector2 sprPos = newRelative + targetSpr;
-
-                float rawLatitudeIndex = (sprPos.x - Mathf.PI / 2) / 6.2831855f * 200;
-                int latitudeIndex = Mathf.FloorToInt(Mathf.Max(0f, Mathf.Abs(rawLatitudeIndex) - 0.1f));
-                int newSegmentCount = PlanetGrid.DetermineLongitudeSegmentCount(latitudeIndex, 200);
-
-                float sizeDeviation = belt.originalSegmentCount / (float)newSegmentCount;
-                if (sizeDeviation > currentMaxWidth)
-                    currentMaxWidth = sizeDeviation;
-
-                if (useExperimentalWidthFix && sizeDeviation < lastMaxWidth)
-                    sizeDeviation = lastMaxWidth;
-
-                sprPos = new Vector2(newRelative.x, newRelative.y * sizeDeviation) + targetSpr;
-
-                Vector3 absoluteBeltPos = sprPos.SnapToGrid(belt.altitude * 1.3333333f / 2);
-
-                // belts have always 0 yaw
-                Quaternion absoluteBeltRot = Maths.SphericalRotation(absoluteBeltPos, 0f);
-
-                BuildPreview bp = BuildPreview.CreateSingle(belt.itemProto, belt.itemProto.prefabDesc, true);
-                bp.ResetInfos();
-                bp.desc = belt.itemProto.prefabDesc;
-                bp.item = belt.itemProto;
-
-                bp.lpos = absoluteBeltPos;
-                bp.lrot = absoluteBeltRot;
-                bp.outputToSlot = -1;
-                bp.outputFromSlot = 0;
-
-                bp.inputFromSlot = -1;
-                bp.inputToSlot = 1;
-
-                bp.outputOffset = 0;
-                bp.inputOffset = 0;
+                PasteBelt(i, data.copiedBelts[i], targetSpr, yaw);
+            }
 
-                Pose pose = new Pose(absoluteBeltPos, absoluteBeltRot);
+            if (pasteInserters && data.copiedInserters.Count > 16)
+            {
 
-                int objId = InserterPoses.AddOverride(pose, belt.itemProto);
+                var maxThreads = 3;
+                var runningThreads = maxThreads;
+                var next = -1;
+                ManualResetEvent done = new ManualResetEvent(false);
 
-                pastedEntities.Add(belt.originalId, new PastedEntity()
+                for (int i = 0; i < maxThreads; i++)
                 {
-                    type = EPastedType.belt,
-                    index = i,
-                    pose = pose,
-                    objId = objId,
-                    buildPreview = bp,
-                });
+                    ThreadPool.QueueUserWorkItem(_ =>
+                    {
+                        int item;
+                        try
+                        {
+                            PlayerAction_Build ab = BuildLogic.ClonePlayerAction_Build(actionBuild);
+                            while ((item = Interlocked.Increment(ref next)) < data.copiedInserters.Count)
```

Comment:
So, `next` is a shared counter between the 3 threads? This makes more of a case for Parallel.For. In example, you could build 3 separate lists, 1 in each of the slice and then merge them at the end.

Basically, this version is a bit more optimized for dispersed sets of data, such as, each item taking a different amount of time. This is okay, but I just want to confirm that I understood this.